feat: add VR water shader

alandtse opened the pull request:
Enables water blending, parallax, etc.
Awaiting discord testing.
https://discord.com/channels/1080142797870485606/1082470343119216751/1224281708590010432

alandtse committed:
feat: add VR water shader

NOTE: There are still bugs.

doodlum committed:
fix: water depth in VR

alandtse committed:
style: 🎨 apply clang-format changes

alandtse committed:
chore: add convertostereosp to vr.hlsl

alandtse committed:
chore: use convertostereosp in getdepth

This is still broken as water at Lake Illinata
`coc guardianstones` is fully transparent. But at
least it doesn't move with HMD.

alandtse committed:
Revert "fix: water depth in VR"

This reverts commit c31be8767f7bf2da695c6af064dabbaafa3bc5d0.

alandtse committed:
chore: remove unnecessary conversions

alandtse commented on package/Shaders/Water.hlsl:
```diff
@@ -633,16 +732,17 @@ PS_OUTPUT main(PS_INPUT input)
 
 	float3 depthAdjustedViewDirection = -viewDirection * depthMul;
-	float viewSurfaceAngle = dot(depthAdjustedViewDirection, ReflectPlane.xyz);
+	float viewSurfaceAngle = dot(depthAdjustedViewDirection, ReflectPlane[eyeIndex].xyz);
```
This and the next line appear to be where this will break. If these are set to a static float value, then the water depth will be wrong, but it will not swim with the headset.

alandtse replied:
ReflectPlane[eyeIndex] is fine. Both 0 and 1 have the exact same value.

alandtse replied:
Fixed, `depthMul` needs to be calculated differently in VR.

alandtse committed:
fix: fix VR water

alandtse committed:
refactor: clarify VR variable names

alandtse committed:
feat: enable waterblending

alandtse committed:
fix: fix compilation failures in compute shaders

Moving water specific function to water.hlsl.

alandtse committed:
refactor: simplify redundant code

alandtse committed:
feat: enable waterparallax for VR

kinobe80 commented:
Noticable flickering of the Water Caustics mod on certain types of water surface. Updated Adress lib. After more diggin around it seems watercaustics.dds is acting weird too but just with vanilla water. seem out of sync or too fast for vr movement speed.

Animation Issues noticed with Vanilla water, Natural Waterfalls water, Simplitcty of sea and more: 
1 Vanilla + Watercuastics.dds = out of sync or too fast Its very hard to notice when in vanilla on lakes and rivers due to the noisy vanilla water and riverground texture with all the rocks, use a brighter flatter riverbededge texture mod. Watercaustics.dds is moving super fast, faster than the river, speeds up even more when walking near a waters edge. Its affected by my Vr movement speed setting from Vr options too where when i lower the setting in skyrimvr the speed of the flow will change but even at slowest it seems watercaustic is moving too fast, two times too fast it seems. The texture appears smaller in the stack amount too. The rings are smaller. Its also affected by the water mod as i noticed when changing water mods the animation syncs as to what im used to from before. 

2 Flickering of water surface inside cities, outside too where a nif is used to build the landscape, like any tundra pond. Vanilla setup. 
heres a screenshot of a random place where it flickers. 
![ScreenShot1](https://github.com/doodlum/skyrim-community-shaders/assets/165723277/8d81b3fe-117a-4bf6-8095-a2cba55b9caa) Seems to be wherever theres water color touching watercaustics.dds there it will flicker. The flickering outside on lakes and rivers is not present with vanilla weather and water , however see next issue.

Issues noticed Natural Waterfalls water, Simplicity of sea, which have esp that changes water colors:
1. Same thing happens as with vanilla where inside city waters and pond waters outside, mesh based, will flicker.
I assume its cause all of that little water surface is using some specific water color or the mod is tryin to apply watercaustics.dds onto the mesh.
2. Edge of rivers and lakes will flicker where the water mod changes the edge of the water with different colors. The water depth color doesnt flicker. Just the edge or close to edge edits so  basically wherever color changes of water touches watercaustics.dds
![ScreenShot17](https://github.com/doodlum/skyrim-community-shaders/assets/165723277/b99fe4ce-4333-4e6b-a194-6857ba3798b1)

alandtse commented:
Thanks. Some inline thoughts.

> Noticable flickering of the Water Caustics mod on certain types of water surface. Updated Adress lib. After more diggin around it seems watercaustics.dds is acting weird too but just with vanilla water. seem out of sync or too fast for vr movement speed.
> 
> * Animation Issues noticed with Vanilla water, Natural Waterfalls water, Simplitcty of sea and more:
>   1 Vanilla + Watercuastics.dds = out of sync or too fast Its very hard to notice when in vanilla on lakes and rivers due to the noisy vanilla water and riverground texture with all the rocks, use a brighter flatter riverbededge texture mod. Watercaustics.dds is moving super fast, faster than the river, speeds up even more when walking near a waters edge. Its affected by my Vr movement speed setting from Vr options too where when i lower the setting in skyrimvr the speed of the flow will change but even at slowest it seems watercaustic is moving too fast, two times too fast it seems. The texture appears smaller in the stack amount too. The rings are smaller. Its also affected by the water mod as i noticed when changing water mods the animation syncs as to what im used to from before.
> 
I'll look a this. I don't believe I modified the speed of the caustic as it's the same, but i'll check.

> 2 Flickering of water surface inside cities, outside too where a nif is used to build the landscape, like any tundra pond. Vanilla setup. heres a screenshot of a random place where it flickers. ![ScreenShot1](https://private-user-images.githubusercontent.com/165723277/318570943-8d81b3fe-117a-4bf6-8095-a2cba55b9caa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIwMzg1NjIsIm5iZiI6MTcxMjAzODI2MiwicGF0aCI6Ii8xNjU3MjMyNzcvMzE4NTcwOTQzLThkODFiM2ZlLTExN2EtNGJmNi04MDk1LWEyY2JhNTViOWNhYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQwMlQwNjExMDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02Nzc0NDU2YzAzMThjOWJhZmVjNmY2NzJmZmVmM2FhMjU2YTdlYWJkYWY5MTA5Mjg5ZjlkNjZkMDc0ZmMwM2Q0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.ONcCjtGHegnjEjJvi50FlVY9FpIuv_56JsNEF9mul8A) Seems to be visible flicker wherever theres water but no water color , like near shores or an object, but watercaustics.dds rendered over it will flicker. The flickering outside on lakes and rivers is not present with vanilla weather and water , however see next issue.
> 
> * Issues noticed Natural Waterfalls water, Simplicity of sea, which have esp that changes water colors:
> 
> 1. Same thing happens as with vanilla where inside city waters and pond waters outside, mesh based, will flicker.
>    I assume its cause all of that little water surface is using some specific non set water color and mod is tryin to apply watercaustics.dds onto the mesh.
> 2. Edge of rivers and lakes will flicker where the water mod changes the edge of the water with no colors. The water depth color doesnt flicker. Just the edge or close to edge edits.
>    ![ScreenShot17](https://private-user-images.githubusercontent.com/165723277/318592245-b99fe4ce-4333-4e6b-a194-6857ba3798b1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIwMzg1NjIsIm5iZiI6MTcxMjAzODI2MiwicGF0aCI6Ii8xNjU3MjMyNzcvMzE4NTkyMjQ1LWI5OWZlNGNlLTQzMzMtNGU2Yi1hMTk0LTY4NTdiYTM3OThiMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQwMlQwNjExMDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jMjRlMGQ3OTRlNzZjZTg1OTMyMDI1YzdlMjY0YThjZTIyOTA1ZTJiOTliOGMyMTBlYzk3MmYwN2Q5MTFmNjg1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.cyyW0GQcS1x3egZdOf9yzPDkDZ6Po1kV-59hOs3In5s)

I feel like without video its hard to know the issue. Again, easy instructions to reproduce will help. Basically vanilla, coc here, walk here. Hell maybe minimal save files would work. I may install 1 mod or 2 max to try to match.

> 3. Installing Water for ENB with Vanilla colors = no river no lake flicker. city water flickers. Installing Water for ENB with other color options = rivers and lakes flicker again.
> 
> * Water reflection reflects my weapons too. Perhaps a limit of sss.
>   ![ScreenShot20](https://private-user-images.githubusercontent.com/165723277/318594600-066dd204-ac89-4411-87e4-6214296ae60a.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIwMzg1NjIsIm5iZiI6MTcxMjAzODI2MiwicGF0aCI6Ii8xNjU3MjMyNzcvMzE4NTk0NjAwLTA2NmRkMjA0LWFjODktNDQxMS04N2U0LTYyMTQyOTZhZTYwYS5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQwMlQwNjExMDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZjMxYmZjODgxNmE5ZTgxNWRmZjc1ZjAxYWFjM2FiZWI3NmIxOTdmZjdkYjVmMjJjMmU5NGQ4OTk2MjQ4ZWU4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.uKSS702derWvZYETedXbydfuOSh40a3u7kOv3LTt_fw)
> * Does not like Fixed Fovieated rendering in Openxr. Notice the white outlines on the grass in the right side. Its never that bright in older cs versions.
>   ![ScreenShot175](https://private-user-images.githubusercontent.com/165723277/318610801-c39dd350-3154-4882-a5ec-71fbdf1a3ff0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIwMzg1NjIsIm5iZiI6MTcxMjAzODI2MiwicGF0aCI6Ii8xNjU3MjMyNzcvMzE4NjEwODAxLWMzOWRkMzUwLTMxNTQtNDg4Mi1hNWVjLTcxZmJkZjFhM2ZmMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQwMlQwNjExMDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03MzVkYmExNjQyYjUxYTM1ZDgyMGM0N2Q5MWE4YzFmMGZiYjFkMWUxOGI5ODNiMzRlMTY3YTFkNjM4OTYwYzQ5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.h3wICQvsqWDG9aFlHZcoKJbqUaaCF-8r_9hgeDeDYss)

If it's SSR, then I'll see if this is a flat problem. Maybe we need to apply a stencil. @FlayaN was this the issue you all saw in SSS?

kinobe80 commented:
Ok heres some videos.

vanilla water animation too fast, out of sync. not the usual caustics movement we have in nexus release.
Its like this everywhere but for example in the video. fast travel to Ilinalta's Deep and look at water edge or around riverwood bridge:
https://streamable.com/fu3sfe 
https://streamable.com/rh5n2a

vanilla water in cities flickers everywhere that there is water. for example whiterun, port to Dragons reach and walk down the stairs.
https://streamable.com/pln8yb

Simplicity of water mod (and any other water mod) in cities and tundra ponds flickers like this. for example whiterun, port to Dragons reach and walk down the stairs.
https://streamable.com/1atn0q

In all rivers and lakes for example around Riverwood bridge in the video, or fast travel to Ilinalta's Deep and look at water edge. This is with Simplicity of waters, but its like this with any water mod i tested like 3-4. Whenever the water mod would change vanilla water colors this happens: 
https://streamable.com/it1smt
https://streamable.com/s9m1tp
https://streamable.com/mkpfu8

If the mod is NOT an esp with color changes but just a retexture then the edge will look like the first issue with the animation.

FlayaN commented:
> @FlayaN was this the issue you all saw in SSS?

No not really, the issue I saw with Screen space shadows in VR without the stencil were rays sampling the border outside the eyes. The effect from that were shadows showing up in some places where they shouldn't. For example arms and when looking straight down sometimes. But it was pretty obvious on the outline that it came from the VR borders

alandtse committed:
fix: fix caustic flickering

alandtse committed:
feat: enable LLF for water shader VR

FlayaN commented on src/Features/LightLimitFix.cpp:
```diff
@@ -1007,7 +1007,7 @@ bool LightLimitFix::HasShaderDefine(RE::BSShader::Type shaderType)
 	case RE::BSShader::Type::Effect:
 	case RE::BSShader::Type::Water:
-		return !REL::Module::IsVR();
+		return true;
```
Maybe move up `case RE::BSShader::Type::Water:` to return true cases above instead? Otherwise we have 2 groups both returning true and also effect returning true for VR.

Like:
```cpp
switch (shaderType) {
	case RE::BSShader::Type::Lighting:
	case RE::BSShader::Type::Grass:
	case RE::BSShader::Type::Water:
		return true;
	case RE::BSShader::Type::Effect:
		return !REL::Module::IsVR();
	default:
		return false;
	}
```

alandtse committed:
fix: fix color bleeding from player character

alandtse committed:
fix: fix case grouping for llf shaders

FlayaN commented on package/Shaders/Water.hlsl:
```diff
@@ -463,7 +581,12 @@ float3 GetWaterSpecularColor(PS_INPUT input, float3 normal, float3 viewDirection
-			reflectionColor = lerp(dynamicCubemap.xyz, CubeMapTex.SampleLevel(CubeMapSampler, cubemapUV, 0).xyz, saturate(length(input.WPosition.xyz) * 0.0001));
+			reflectionColor =
+#			if defined(VR)  // use stencil to ignore player character
+				GetStencil(cubemapUV) == 0 ? CubeMapTex.SampleLevel(CubeMapSampler, cubemapUV, 0).xyz :
```
`cubemapUV` is a float3, add `.xy` here to fix implicit truncation

FlayaN commented on package/Shaders/Water.hlsl:
```diff
@@ -279,12 +348,61 @@ cbuffer PerMaterial : register(b1)
+	return DepthTex.Load(int3(uv * lightingData[0].BufferDim * DynamicResolutionParams1.xy, 0)).g;
+}
+
+float GetStencil(float2 uv, uint a_eyeIndex)
```
Only above func used, remove this one

FlayaN commented on package/Shaders/Water.hlsl:
```diff
@@ -463,7 +581,12 @@ float3 GetWaterSpecularColor(PS_INPUT input, float3 normal, float3 viewDirection
+				GetStencil(cubemapUV) == 0 ? CubeMapTex.SampleLevel(CubeMapSampler, cubemapUV, 0).xyz :
+#			endif
+											 lerp(dynamicCubemap.xyz, CubeMapTex.SampleLevel(CubeMapSampler, cubemapUV, 0).xyz, saturate(length(input.WPosition.xyz) * 0.0001));
+			;
```
Remove extra `;`

alandtse commented:
Ok, new report. Moving headset up and down results in caustics moving up and down. Moving back to draft. I'll probably not work on this for a bit so will let someone else finish it.

alandtse committed:
chore: cleanup unused code

alandtse committed:
fix: disable ssr reflection logic in VR

In VR, the texture is all black.

alandtse committed:
fix: fix implicit truncation in ComputeWaterCaustics

alandtse committed:
chore: remove redundant functions

The #ifdef Water complicates the code more than is probably needed.

alandtse committed:
fix: disable bad caustics
